lib/elixir/lib/macro/env.ex

```diff
@@ -580,13 +591,19 @@ defmodule Macro.Env do
         trace = Keyword.get(opts, :trace, true)
         module = env.module
 
+        # When local_for_callback is provided, we don't need to pass module macros as extra
+        # because the callback will handle local macro resolution
         extra =
-          case allow_locals and function_exported?(module, :__info__, 1) do
-            true -> [{module, module.__info__(:macros)}]
-            false -> []
+          if Keyword.has_key?(opts, :local_for_callback) do
+            []
+          else
+            case allow_locals and function_exported?(module, :__info__, 1) do
+              true -> [{module, module.__info__(:macros)}]
+              false -> []
+            end
           end
 
-        case :elixir_dispatch.expand_import(meta, name, arity, env, extra, allow_locals, trace) do
+        case :elixir_dispatch.expand_import(meta, name, arity, env, extra, allow_locals, trace, local_for_callback) do
```

Comment:
Instead of adding a new option, we can augment `allow_locals`. If it is false, it is disabled, if it is true, it is the default implementation (`elixir_locals`), otherwise it is a custom function.

Reply:
Done on the erlang side. Do you also want that on the elixir API side?